### PR TITLE
[Merged by Bors] - feat: port Data.Stream.Defs

### PR DESCRIPTION
**Pull request by ericrbg**
Mathlib SHA: 39af7d3bf61a98e928812dbc3e16f4ea8b795ca3

Diff view: https://github.com/leanprover-community/mathlib4/compare/7f9d342...stream

### CODE REVIEW

**Comment by ericrbg** [comment]
I don't know enough about streams to push non-tautological documentation, so if people want feel free to push some

File: Mathlib/Data/Stream/Defs.lean [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+variable {α : Type u} {β : Type v} {δ : Type w}
+
+/-- Prepend an element to a stream. -/
+def cons (a : α) (s : Stream' α) : Stream' α
+  | 0 => a
+  | n + 1 => s n
+#align stream.cons Stream'.cons
+
+-- mathport name: stream.cons
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Stream/Defs.lean:34]:
```suggestion
```

File: Mathlib/Data/Stream/Defs.lean [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+def unfolds (g : α → β) (f : α → α) (a : α) : Stream' β :=
+  corec g f a
+#align stream.unfolds Stream'.unfolds
+
+/-- Interleave two streams. -/
+def interleave (s₁ s₂ : Stream' α) : Stream' α :=
+  corecOn (s₁, s₂) (fun ⟨s₁, _⟩ => head s₁) fun ⟨s₁, s₂⟩ => (s₂, tail s₁)
+#align stream.interleave Stream'.interleave
+
+-- mathport name: «expr ⋈ »
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Stream/Defs.lean:119]:
```suggestion
```

File: Mathlib/Data/Stream/Defs.lean [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+  even (tail s)
+#align stream.odd Stream'.odd
+
+/-- Append a stream to a list. -/
+def appendStream' : List α → Stream' α → Stream' α
+  | [], s => s
+  | List.cons a l, s => a::appendStream' l s
+#align stream.append_stream Stream'.appendStream'
+
+-- mathport name: «expr ++ₛ »
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Stream/Defs.lean:138]:
```suggestion
```

File: Mathlib/Data/Stream/Defs.lean [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+/-- A constant stream, same as `Stream'.const`. -/
+def pure (a : α) : Stream' α :=
+  const a
+#align stream.pure Stream'.pure
+
+/-- Given a stream of functions and a stream of values, apply `n`-th function to `n`-th value. -/
+def apply (f : Stream' (α → β)) (s : Stream' α) : Stream' β := fun n => (nth f n) (nth s n)
+#align stream.apply Stream'.apply
+
+-- mathport name: «expr ⊛ »
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Stream/Defs.lean:190]:
```suggestion
```

**Review by Ruben-VandeVelde — APPROVED** [review]
lgtm

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**